server/src/main/java/org/elasticsearch/search/aggregations/InternalOrder.java

```diff
@@ -489,14 +494,14 @@ public static BucketOrder readOrder(StreamInput in) throws IOException {
                 case Aggregation.ID:
                     boolean asc = in.readBoolean();
                     String key = in.readString();
-                    return new Aggregation(key, asc);
+                    return bucketOrderDeduplicator.deduplicate(new Aggregation(key, asc));
                 case CompoundOrder.ID:
                     int size = in.readVInt();
                     List<BucketOrder> compoundOrder = new ArrayList<>(size);
                     for (int i = 0; i < size; i++) {
                         compoundOrder.add(Streams.readOrder(in));
                     }
-                    return new CompoundOrder(compoundOrder, false);
+                    return bucketOrderDeduplicator.deduplicate(new CompoundOrder(compoundOrder, false));
```

Comment:
ESQL uses a wrapper around the `StreamInput` that keeps the cache in a regular old variable rather than a static. I'd prefer that if we can manage it.

Reply:
I have moved it as a wrapper of StreamInput by (ab)using the fact that aggregations are deserialize using DelayableWritable. I have to introduce an interface so we can deduplicate when it is found.